pkg/controller/import-controller.go

```diff
@@ -44,315 +51,532 @@ const (
+				return reconcile.Result{}, err
 			}
+			return reconcile.Result{}, nil
 		}
 
-		if pod.Status.Phase == v1.PodSucceeded || scratchExitCode {
-			dReq := podDeleteRequest{
-				namespace: pod.Namespace,
-				podName:   pod.Name,
-				podLister: ic.Controller.podLister,
-				k8sClient: ic.Controller.clientset,
-			}
-			// just use defer here so we make sure our pvc updates get written prior to actual deletion
-			defer deletePod(dReq)
+		// Pod exists, we need to update the PVC status.
+		if err := r.updatePvcFromPod(pvc, pod); err != nil {
+			return reconcile.Result{}, err
 		}
-		if pod.Status.Phase == v1.PodPending && ic.requiresScratchSpace(pvc) {
-			err = ic.createScratchPvcForPod(pvc, pod)
-			if err != nil {
-				return err
-			}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ImportReconciler) updatePvcFromPod(pvc *corev1.PersistentVolumeClaim, pod *corev1.Pod) error {
+	// Keep a copy of the original for comparison later.
+	currentPvcCopy := pvc.DeepCopyObject()
+
+	r.Log.V(1).Info("Updating PVC from pod")
+	anno := pvc.GetAnnotations()
+	scratchExitCode := false
+	if pod.Status.ContainerStatuses != nil && pod.Status.ContainerStatuses[0].LastTerminationState.Terminated != nil &&
+		pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode > 0 {
+		r.Log.Info("Pod termination code", "pod.Name", pod.Name, "ExitCode", pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode)
+		if pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.ExitCode == common.ScratchSpaceNeededExitCode {
+			r.Log.V(1).Info("Pod requires scratch space, terminating pod, and restarting with scratch space", "pod.Name", pod.Name)
+			scratchExitCode = true
+			anno[AnnRequiresScratch] = "true"
+		} else {
+			r.recorder.Event(pvc, corev1.EventTypeWarning, ErrImportFailedPVC, pod.Status.ContainerStatuses[0].LastTerminationState.Terminated.Message)
 		}
 	}
 
+	anno[AnnImportPod] = string(pod.Name)
+	if !scratchExitCode {
+		r.updatePvcPhaseFromPod(anno, pod)
+	}
+
+	// Check if the POD is waiting for scratch space, if so create some.
+	if pod.Status.Phase == corev1.PodPending && r.requiresScratchSpace(pvc) {
+		if err := r.createScratchPvcForPod(pvc, pod); err != nil {
+			return err
+		}
+	}
 	if !checkIfLabelExists(pvc, common.CDILabelKey, common.CDILabelValue) {
-		lab = map[string]string{common.CDILabelKey: common.CDILabelValue}
+		if pvc.GetLabels() == nil {
+			pvc.SetLabels(make(map[string]string, 0))
+		}
+		pvc.GetLabels()[common.CDILabelKey] = common.CDILabelValue
 	}
 
-	pvc, err = updatePVC(ic.clientset, pvc, anno, lab)
-	if err != nil {
-		return errors.WithMessage(err, "could not update pvc %q annotation and/or label")
+	if !reflect.DeepEqual(currentPvcCopy, pvc) {
+		if err := r.updatePVC(pvc); err != nil {
+			return err
+		}
+		r.Log.V(1).Info("Updated PVC", "pvc.anno.Phase", anno[AnnPodPhase])
+	}
+
+	if pvc.GetAnnotations()[AnnPodPhase] == string(corev1.PodSucceeded) || scratchExitCode {
```

Comment:
maybe call `isPVCComplete()` instead of `pvc.GetAnnotations()[AnnPodPhase] == string(corev1.PodSucceeded)`

Reply:
done